packages/swing-store/src/snapStore.js

```diff
@@ -217,63 +210,117 @@ export function makeSnapStore(
+        cleanup.push(() => {
+          snapReader.destroy();
+        });
+
+        // Create a file for the compressed snapshot in-place
+        // to be atomically renamed once we know its name
+        // from the hash of raw snapshot contents.
+        const {
+          path: tmpGzPath,
+          fd: tmpGzFd,
+          cleanup: tmpGzCleanup,
+        } = await ptmpFile({ tmpdir: root });
+        cleanup.push(tmpGzCleanup);
+        const gzWriter = createWriteStream(noPath, {
+          fd: tmpGzFd,
+          autoClose: false,
+        });
+        cleanup.push(() => gzWriter.close());
+
+        await Promise.all([fsStreamReady(snapReader), fsStreamReady(gzWriter)]);
+
+        const hashStream = createHash('sha256');
+        const gzip = createGzip();
+
+        const { result: hash, duration: compressSeconds } =
+          await measureSeconds(async () => {
+            snapReader.pipe(hashStream);
+            snapReader.pipe(gzip).pipe(gzWriter);
+
+            await Promise.all([finished(snapReader), finished(gzWriter)]);
+
+            const h = hashStream.digest('hex');
+            await pfsync(tmpGzFd);
+
+            const tmpGzClose = cleanup.pop();
+            tmpGzClose();
+
+            return h;
+          });
+
+        toDelete.delete(hash);
+        const baseName = baseNameFromHash(hash);
+        const filePath = pathResolve(root, baseName);
+        const infoBase = {
+          filePath,
+          hash,
+          rawByteCount,
+          rawSaveSeconds,
+          compressSeconds,
+        };
+        const fileStat = await stat(filePath).catch(err => {
+          if (err.code !== 'ENOENT') {
+            throw err;
+          }
+        });
+        if (fileStat) {
+          const { size: compressedByteCount } = fileStat;
+          return freeze({
+            ...infoBase,
+            newFile: false,
+            compressedByteCount,
+          });
         }
-        throw e;
-      });
-      if (fileStat) {
-        const { size: compressedByteCount } = fileStat;
+
+        // Atomically rename.
+        await rename(tmpGzPath, filePath);
+        const { size: compressedByteCount } = await stat(filePath);
         return freeze({
           ...infoBase,
-          newFile: false,
-          compressSeconds: 0,
+          newFile: true,
           compressedByteCount,
         });
-      }
-      const { duration: compressSeconds } = await measureSeconds(() =>
-        atomicWriteInRoot(baseName, tmpGzPath =>
-          filter(tmpSnapPath, createGzip(), tmpGzPath, { flush: true }),
-        ),
-      );
-      const { size: compressedByteCount } = await stat(filePath);
-      return freeze({
-        ...infoBase,
-        newFile: true,
-        compressSeconds,
-        compressedByteCount,
-      });
-    }, 'save-raw');
+      },
+      async () => {
+        await PromiseAllOrErrors(
+          cleanup.reverse().map(fn => Promise.resolve().then(() => fn())),
+        );
```

Comment:
Thinking about this more, shouldn't we wait for previous steps to be done to continue with the next cleanup step? Disposal of resource is hard!

Reply:
Not an actual problem since only the last cleanup (first queued) is actually async. Let's follow-up to clean this up though.